### ISSUE TITLE
libtorch C++ setting gradient to 0 is incorrectly detected as violating const correctness

### ISSUE THREAD

**Issue by CBSkarmory** [pytorch/pytorch#41264]
## 🐛 Bug

<!-- A clear and concise description of what the bug is. -->
compiling incorrectly fails with the error `error: passing ‘at::Tensor’ as ‘this’ argument discards qualifiers [-fpermissive]` (aka const correctness violated)

This follows the example from the docs (https://pytorch.org/cppdocs/api/classat_1_1_tensor.html#_CPPv4N2at6Tensor4gradEv)

> `Tensor &grad()`
> Return a mutable reference to the gradient. 
> This is conventionally used as `t.grad() = x` to set a gradient to a completely new tensor.

> `void backward(...)`
> ...
> This function accumulates gradients in the leaves - you might need to zero them before calling it.

but fails to compile

## To Reproduce


1. call backward()
2. set the grad to zero
3. compile

<!-- If you have a code sample, error messages, stack traces, please provide it here as well -->
```
...error: passing ‘at::Tensor’ as ‘this’ argument discards qualifiers [-fpermissive]
     z_grad = 0;
              ^
```
```
#include <torch/torch.h>                                                                                                                                                                                           
#include <iostream>                                                                                                                                                                                                
#include <vector>                                                                                                                                                                                                  
                                                                                                                                                                                                                   torch::Device get_device()                                                                                                                                                                                         
{                                                                                                                                                                                                                  
    torch::Device device = torch::kCPU;                                                                                                                                                                            
    if (torch::cuda::is_available()) {                                                                                                                                                                                     std::cout << "CUDA is available! Training on GPU." << std::endl;                                                                                                                                           
        device = torch::kCUDA;                                                                                                                                                                                     
    }                                                                                                                                                                                                              
    return device;                                                                                                                                                                                                 
}                                                                                                                                                                                                                  
                                                                                                                                                                                                                   
int main() {                                                                                                                                                                                                       
    using torch::Tensor;                                                                                                                                                                                           
    auto device = get_device();                                                                                                                                                                                    
    auto options =                                                                                                                                                                                                 
        torch::TensorOptions()                                                                                                                                                                                     
        .dtype(torch::kFloat32)                                                                                                                                                                                    
        .layout(torch::kStrided)                                                                                                                                                                                   
        .device(device)                                                                                                                                                                                            
        .requires_grad(true);                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                                                              
    std::vector<float> z_vec = {1,2,3};                                                                                                                                                                                                                                                                   
    Tensor z = torch::from_blob(z_vec.data(), {z_vec.size()}).to(device);                                                                                                                                          
                                                                                                                                                                                                                   
    Tensor z_times_2 = 2 * z;                                                                                                                                                                                      
    z_times_2.backward({}, true); // keep graph                                                                                                                                                                    
    Tensor& z_grad = z.grad(); // mutable, non-const ref
    std::cout << z << std::endl;                                                                                                                                                                                   
    std::cout << z_grad << std::endl;                                                                                                                                                                              
    z_grad = 0;                                                                                                                                                                                                    
    std::cout << z_grad << std::endl;                                                                                                                                                                              
}
```
fails to compile with the error above, identifying `z_grad = 0;` as the problem   

Exactly as used in the docs, `z.grad() = 0` does not work either.

Even with const_cast, it still fails to compile.
`const_cast<Tensor&>(z_grad) = 0; ` or alternatively `Tensor& z_grad = const_cast<Tensor&>(z.grad()); // mutable, non-const ref`

Passing `-fpermissive` to the compiler does allow it to compile, but this should not be necessary

## Expected behavior

It should compile and set the gradient to zero.

## Environment

 - PyTorch Version (e.g., 1.0): libtorch 1.5.1(stable, linux, libtorch, C++/Java, CUDA 10.2) CXX11 ABI
 - OS (e.g., Linux): Linux, Ubuntu 18.04 LTS
 - How you installed PyTorch (`conda`, `pip`, source): the downloads page (https://pytorch.org/get-started/locally/) and linking against it (https://download.pytorch.org/libtorch/cu102/libtorch-cxx11-abi-shared-with-deps-1.5.1.zip)
 - Build command you used (if compiling from source): not compiled from source
 - Python version: irrelevant
 - CUDA/cuDNN version: (probably irrelevant) CUDA 10.2, cuDNN 7.6.5
 - GPU models and configuration: (probably irrelevant) Quadro P1000
 - Additional information: gcc 7.3 (`g++ (Ubuntu 7.3.0-16ubuntu3) 7.3.0`), cmake is pretty much exactly the minimal example here (https://pytorch.org/cppdocs/installing.html) but renamed and without MSVC


cc @ezyang @SsnL @albanD @zou3519 @gqchen @jlin27

**Comment by albanD** [comment]
Hi,

This was changed very recently a accessing this non-const version is not thread safe and using overload here was dangerous.
You can now access it using `.mutable_grad()`.
I am not sure why the doc is not updated though.

**Comment by albanD** [comment]
Hi,

We are working on fixing that in https://github.com/pytorch/pytorch/pull/40887 
You will be able to call an explicit function to get a non-const version.

**Comment by albanD** [comment]
This is now done. You can acces the `.mutable_grad()` to be able to modify it.
